client/src/components/Calendar.jsx

```diff
@@ -0,0 +1,198 @@
+    };
+  }
+
+  previousMonth() {
+    let left = moment(this.state.leftMonth).subtract(1, 'months');
+    let right = moment(this.state.rightMonth).subtract(1, 'months');
+    this.setState({
+      leftMonth: left,
+      rightMonth: right
+    });
+  }
+    
+  nextMonth() {
+    let left = moment(this.state.leftMonth).add(1, 'months');
+    let right = moment(this.state.rightMonth).add(1, 'months');
+    this.setState({
+      leftMonth: left,
+      rightMonth: right
+    });
+  }
+
+  firstDayOfMonth(leftOrRight) {
+    let firstDay = moment(leftOrRight).startOf('month').format('d');
+    return firstDay;
+  }
+
+  render() {
+    const eachDayOfWeek = moment.weekdaysShort().map(day => (
+      <DayOfMonth key={day}>
+        {day.slice(0, -1)}
+      </DayOfMonth>
+    ));
+
+    let leftFiller = [];
+    for (let i = 0; i < this.firstDayOfMonth(this.state.leftMonth); i += 1) {
+      leftFiller.push(
+        <CalendarDay style={{ border: 'none' }}></CalendarDay>
+      );
+    }
+
+    let rightFiller = [];
+    for (let i = 0; i < this.firstDayOfMonth(this.state.rightMonth); i += 1) {
+      rightFiller.push(
+        <CalendarDay style={{ border: 'none' }}></CalendarDay>
+      );
+    }
+
+    let daysInLeftMonth = [];
+    for (let d = 1; d <= this.state.leftMonth.daysInMonth(); d += 1) {
+      daysInLeftMonth.push(
+        <CalendarDay style={{ border: 'solid', borderWidth: 'thin', borderColor: 'grey' }} key={d}>
+          {d}
+        </CalendarDay>
+      );
+    }
+
+    let daysInRightMonth = [];
+    for (let d = 1; d <= this.state.rightMonth.daysInMonth(); d += 1) {
+      daysInRightMonth.push(
+        <CalendarDay style={{ border: 'solid', borderWidth: 'thin', borderColor: 'grey' }} key={d}>
+          {d}
+        </CalendarDay>
+      );
+    }
+
+    let totalLeftCalendar = [...leftFiller, ...daysInLeftMonth];
+    let totalRightCalendar = [...rightFiller, ...daysInRightMonth];
+    let rowsOfDaysLeft = [];
+    let rowsOfDaysRight = [];
+    let daysPerEachWeekLeft = [];
+    let daysPerEachWeekRight = [];
+
+    totalLeftCalendar.forEach((row, i) => {
+      if (i % 7 !== 0) {
+        daysPerEachWeekLeft.push(row);
+      } else {
+        rowsOfDaysLeft.push(daysPerEachWeekLeft);
+        daysPerEachWeekLeft = [];
+        daysPerEachWeekLeft.push(row);
+      }
+      if (i === totalLeftCalendar.length - 1) {
+        rowsOfDaysLeft.push(daysPerEachWeekLeft);
+      }
+    });
+
+    totalRightCalendar.forEach((row, i) => {
+      if (i % 7 !== 0) {
+        daysPerEachWeekRight.push(row);
+      } else {
+        rowsOfDaysRight.push(daysPerEachWeekRight);
+        daysPerEachWeekRight = [];
+        daysPerEachWeekRight.push(row);
+      }
+      if (i === totalRightCalendar.length - 1) {
+        rowsOfDaysRight.push(daysPerEachWeekRight);
+      }
+    });
+
+    let daysinLeftmonth = rowsOfDaysLeft.map(d => <tr>{d}</tr>);
+    let daysinRightmonth = rowsOfDaysRight.map(d => <tr>{d}</tr>);
```

Comment:
This whole `render` function is getting a little bit long and hard to follow the logic and tasks done in here.  It might not be a bad idea to have one or more functions to put this logic there. Maybe a simple `preprocess()` function.

Also, another thing to consider is `render` function in React can be called several times.  So, while performance might not be a concern to us right now, it would be a healthy practice to keep render as lightweight as possible, by shifting some one-time-only calculations to the constructor or other methods in the lifecycle that are less frequent than `render` (e.g. `componentDidMount`.

Reply:
Thank you for the thorough review. I will implement these changes!